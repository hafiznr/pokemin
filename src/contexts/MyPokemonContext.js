import React, { createContext, useState } from 'react';

export const MyPokemonContext = createContext();

export const MyPokemonContextProvider = ({ children, initialData }) => {
  const [myPokemonData, setMyPokemonData] = useState(JSON.parse(initialData) || []);

  const addPokemon = (data) => {
    const newPokemonData = [...myPokemonData, data];

    setMyPokemonData(newPokemonData);

    localStorage.setItem('mypokemon', JSON.stringify(newPokemonData));
  }

  const removePokemon = (nickname) => {
    const newPokemonData = myPokemonData
      .filter(pokemon => pokemon.nickname !== nickname);

    setMyPokemonData(newPokemonData);

    localStorage.setItem('mypokemon', JSON.stringify(newPokemonData));
  }

  return (
    <MyPokemonContext.Provider 
      value={{ myPokemonData, addPokemon, removePokemon }}
    >
      {children}
    </MyPokemonContext.Provider>
  )
}
