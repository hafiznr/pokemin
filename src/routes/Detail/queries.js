import { gql } from '@apollo/client';

export const GET_POKEMON = gql`
  query pokemon($name: String!) {
    pokemon(name: $name) {
      abilities {
        ability {
          url
          name
        }
        is_hidden
        slot
      }
      base_experience
      height
      id
      moves {
        move {
          url
          name
        }
      }
      name
      species {
        url
        name
      }
      sprites {
        back_default
        back_female
        back_shiny
        back_shiny_female
        front_default
        front_female
        front_shiny
        front_shiny_female
      }
      types {
        slot
        type {
          url
          name
        }
      }
      weight
      status
      message
    }
  }
`;