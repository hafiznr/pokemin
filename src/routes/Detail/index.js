import { useContext, useState } from 'react';
import { useParams } from 'react-router';
import { useQuery } from '@apollo/client';
import { 
  Button,
  Skeleton,
  Box
} from '@chakra-ui/react';

// queries
import { GET_POKEMON } from './queries';

// components
import SuccessModal from './components/SuccessModal';
import FailedModal from './components/FailedModal';
import Moves from './components/Moves';
import Summary from './components/Summary';

// utils
import { capitalizeFirstLetter } from '../../utils';

// context
import { MyPokemonContext } from '../../contexts/MyPokemonContext';

const Detail = () => {
  const { addPokemon } = useContext(MyPokemonContext);

  const [showModal, setShowModal] = useState('');

  const { pokemonId } = useParams();

  const { loading, error, data: pokemonData } = useQuery(GET_POKEMON, {
    variables: {
      name: pokemonId
    }
  });

  const { pokemon } = pokemonData || {};
  const {
    id,  
    moves,
    name = '', 
    sprites,
  } = pokemon || {};

  const capitalizedName = capitalizeFirstLetter(name);

  const handleClickCatch = () => {
    const probability = +process.env.REACT_APP_CATCH_PROBABILITY;
    const isSuccess = Math.random() <= probability;

    if(isSuccess) {
      setShowModal('success');
    } else {
      setShowModal('failed');
    }
  }

  const handleAddPokemon = (nickname) => {
    addPokemon({
      id,
      name,
      image: sprites.front_default,
      nickname
    });

    setShowModal('');
  }

  return (
    <Box className="detail" padding="8px 16px 78px 16px">
      {error && (
        <Box mb="8px" fontSize="xl" fontWeight="bold">
          Something went wrong. Please try again later
        </Box>
      )}

      <Box 
        textAlign="center" 
        fontSize="2xl" 
        fontWeight="bold" 
        mb="8px"
      >
        {loading ? (
          <Skeleton 
            height="27px" 
            mb="8px" 
            data-testid="name-skeleton"
          />
        ) : (
          <div data-testid="capital-name">
            {capitalizedName}
          </div>
        )}
      </Box>

      <Summary loading={loading} data={pokemon} />

      <Moves loading={loading} moves={moves} />

      <Box
        width="100%"
        textAlign="center"
        margin="16px 0"
      >
        {loading ? (
          <Skeleton 
            height="36px" 
            data-testid="catch-button-skeleton"
          />
        ) : (
          <Button 
            size="lg"
            colorScheme="twitter"
            width="100%"
            onClick={handleClickCatch}
            data-testid="catch-button"
          >
            Catch!
          </Button>
        )}
      </Box>

      <SuccessModal
        isOpen={showModal === 'success'}
        pokemon={capitalizedName}
        image={sprites?.front_default}
        onClose={() => setShowModal('')}
        onSubmit={handleAddPokemon}
      />

      <FailedModal
        isOpen={showModal === 'failed'}
        pokemon={capitalizedName}
        image={sprites?.back_default}
        onClose={() => setShowModal('')}
        onRetry={handleClickCatch}
      />
    </Box>
  );
}

export default Detail;
