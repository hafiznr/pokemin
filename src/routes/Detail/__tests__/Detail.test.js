import { MockedProvider } from '@apollo/client/testing';
import { fireEvent, render, screen, act } from '@testing-library/react';
import Detail from '../';
import { MyPokemonContextProvider } from '../../../contexts/MyPokemonContext';
import { BrowserRouter } from 'react-router-dom';
import { GET_POKEMON } from '../queries';

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    pokemonId: "blastoise"
  })
}));

test('renders Detail with loading', () => {
  const mocks = [];

  const { container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <BrowserRouter>
        <MyPokemonContextProvider initialData={null}>
          <Detail />
        </MyPokemonContextProvider>
      </BrowserRouter>
    </MockedProvider>
  );

  expect(container).toMatchSnapshot();
  expect(screen.getByTestId('name-skeleton')).toBeInTheDocument();
  expect(screen.getByTestId('catch-button-skeleton')).toBeInTheDocument();
});

test('renders Detail with results', async () => {
  const mocks = [
    {
      request: {
        query: GET_POKEMON,
        variables: {
          name: 'Blastoise',
        },
      },
      result: {
        data: {
          pokemon: {
            abilities: [],
            base_experience: 239,
            height: 16,
            id: 9,
            message: "",
            moves: [],
            name: "blastoise",
            sprites: {
              back_default: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/9.png",
              front_default: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/front/9.png"
            },
            status: true,
            types: [],
          }
        },
      },
    }
  ];

  const { container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <BrowserRouter>
        <MyPokemonContextProvider initialData={null}>
          <Detail />
        </MyPokemonContextProvider>
      </BrowserRouter>
    </MockedProvider>
  );

  await act(() => new Promise(resolve => setTimeout(resolve, 0)));

  expect(container).toMatchSnapshot();
  expect(screen.getByTestId('capital-name')).toBeInTheDocument();
  expect(screen.getByTestId('catch-button')).toBeInTheDocument();

  process.env.REACT_APP_CATCH_PROBABILITY = "1";
  fireEvent.click(screen.getByTestId('catch-button'));

  process.env.REACT_APP_CATCH_PROBABILITY = "0";
  fireEvent.click(screen.getByTestId('catch-button'));
});