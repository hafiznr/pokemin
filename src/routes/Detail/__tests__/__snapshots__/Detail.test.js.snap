// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Detail with loading 1`] = `
<div>
  <div
    class="detail css-1nsfvvx"
  >
    <div
      class="css-1n890a7"
    >
      <div
        class="chakra-skeleton css-rych3e"
        data-testid="name-skeleton"
      />
    </div>
    <div
      class="css-jxadsl-Summary"
    >
      <div
        class="css-5g37ib-Summary"
      >
        <div
          class="chakra-skeleton css-ibmh67"
        />
      </div>
      <table
        class="chakra-table css-0"
        role="table"
      >
        <tbody
          class="css-0"
        >
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Number
            </td>
            <td
              class="css-0"
              role="gridcell"
            >
              <div
                class="chakra-skeleton css-sf9g0d"
                data-testid="summary-skeleton"
              />
            </td>
          </tr>
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Types
            </td>
            <td
              class="css-0"
              role="gridcell"
            >
              <div
                class="chakra-skeleton css-sf9g0d"
                data-testid="summary-skeleton"
              />
            </td>
          </tr>
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Height
            </td>
            <td
              class="css-0"
              role="gridcell"
            >
              <div
                class="chakra-skeleton css-sf9g0d"
                data-testid="summary-skeleton"
              />
            </td>
          </tr>
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Weight
            </td>
            <td
              class="css-0"
              role="gridcell"
            >
              <div
                class="chakra-skeleton css-sf9g0d"
                data-testid="summary-skeleton"
              />
            </td>
          </tr>
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Abilities
            </td>
            <td
              class="css-0"
              role="gridcell"
            >
              <div
                class="chakra-skeleton css-sf9g0d"
                data-testid="summary-skeleton"
              />
            </td>
          </tr>
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Base exp.
            </td>
            <td
              class="css-0"
              role="gridcell"
            >
              <div
                class="chakra-skeleton css-sf9g0d"
                data-testid="summary-skeleton"
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div>
      <div
        class="css-bn3r0t"
      >
        Moves
      </div>
      <div
        class="chakra-skeleton css-yy4ngl"
        data-testid="moves-skeleton"
      />
    </div>
    <div
      class="css-cmyg6m"
    >
      <div
        class="chakra-skeleton css-qouu9h"
        data-testid="catch-button-skeleton"
      />
    </div>
  </div>
</div>
`;

exports[`renders Detail with results 1`] = `
<div>
  <div
    class="detail css-1nsfvvx"
  >
    <div
      class="css-1yz1o5g"
    >
      Something went wrong. Please try again later
    </div>
    <div
      class="css-1n890a7"
    >
      <div
        data-testid="capital-name"
      />
    </div>
    <div
      class="css-jxadsl-Summary"
    >
      <div
        class="css-5g37ib-Summary"
      >
        <div
          class="chakra-skeleton css-ibmh67"
        />
      </div>
      <table
        class="chakra-table css-0"
        role="table"
      >
        <tbody
          class="css-0"
        >
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Number
            </td>
            <td
              class="css-0"
              role="gridcell"
            />
          </tr>
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Types
            </td>
            <td
              class="css-0"
              role="gridcell"
            />
          </tr>
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Height
            </td>
            <td
              class="css-0"
              role="gridcell"
            />
          </tr>
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Weight
            </td>
            <td
              class="css-0"
              role="gridcell"
            />
          </tr>
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Abilities
            </td>
            <td
              class="css-0"
              role="gridcell"
            >
              <ol
                class="css-1bd4wsw"
                role="list"
              />
            </td>
          </tr>
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-jhl0h9"
              role="gridcell"
            >
              Base exp.
            </td>
            <td
              class="css-0"
              role="gridcell"
            />
          </tr>
        </tbody>
      </table>
    </div>
    <div>
      <div
        class="css-bn3r0t"
      >
        Moves
      </div>
      <div
        class="css-8ehw8a"
        data-testid="moves-container"
      />
    </div>
    <div
      class="css-cmyg6m"
    >
      <button
        class="chakra-button css-13zfu4m"
        data-testid="catch-button"
        type="button"
      >
        Catch!
      </button>
    </div>
  </div>
</div>
`;
